From d3ae7700a6f45a5ac1fefc96484c3d54ca94f4c4 Mon Sep 17 00:00:00 2001
From: Arnavion <me@arnavion.dev>
Date: Fri, 14 Oct 2022 10:43:01 -0700
Subject: [PATCH] Make gdk_wayland_device_set_selection also consider serial
 from touch events.

gdk_wayland_device_set_selection does not have information of which event
(if any) triggered it, so it needs to look up the last serial and use that.
Before this change, it looke up the last serial using
_gdk_wayland_seat_get_implicit_grab_serial, which ignores touch events and
only processes pointer events, so it would only work for devices which have
a mouse. Thus it was broken on devices which only have a touchscreen, like
phones.

This change changes it to use _gdk_wayland_seat_get_last_implicit_grab_serial
which also considers touch events.

Ref: https://gitlab.gnome.org/GNOME/gtk/-/issues/5250
---
 gdk/wayland/gdkdevice-wayland.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gdk/wayland/gdkdevice-wayland.c b/gdk/wayland/gdkdevice-wayland.c
index cd314b6d67..625dda7d32 100644
--- a/gdk/wayland/gdkdevice-wayland.c
+++ b/gdk/wayland/gdkdevice-wayland.c
@@ -5322,7 +5322,7 @@ gdk_wayland_device_set_selection (GdkDevice             *gdk_device,
   g_return_if_fail (GDK_IS_WAYLAND_DEVICE (gdk_device));
 
   seat = GDK_WAYLAND_SEAT (gdk_device_get_seat (gdk_device));
-  serial = _gdk_wayland_seat_get_implicit_grab_serial (GDK_SEAT (seat), NULL);
+  serial = _gdk_wayland_seat_get_last_implicit_grab_serial (GDK_WAYLAND_SEAT (seat), NULL);
   wl_data_device_set_selection (seat->data_device, source, serial);
 }
 
-- 
2.37.3

